src/main/java/com/emc/object/s3/S3SignerV4.java

```diff
@@ -0,0 +1,397 @@
+        try {
+            Date d = sdf.parse(date);
+            sdf.applyPattern(AMZ_DATE_FORMAT_SHORT);
+            shortDate = sdf.format(d);
+            return shortDate;
+        } catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected String getScope(String shortDate, String service) {
+        return shortDate + "/"
+                + S3Constants.AWS_DEFAULT_REGION + "/"
+                + service + "/"
+                + S3Constants.AWS_V4_TERMINATOR;
+    }
+
+    protected String getSignature(String stringToSign, byte[] signingKey) {
+        try {
+            return hexEncode(hmac(S3Constants.HMAC_SHA_256, signingKey, stringToSign));
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to get getSignature");
+        }
+    }
+
+    protected String getServiceType() {
+        return S3Constants.AWS_SERVICE_S3;
+    }
+
+    public URL generatePresignedUrl(PresignedUrlRequest request) {
+        String namespace = request.getNamespace() != null ? request.getNamespace() : s3Config.getNamespace();
+        URI uri = s3Config.resolvePath(request.getPath(), null); // don't care about the query string yet
+        // must construct both the final URL and the resource for signing
+        String resource = "/" + request.getBucketName() + RestUtil.getEncodedPath(uri); // so here we have a uri encoding
+
+        // insert namespace in host
+        if (namespace != null) {
+            if (s3Config.isUseVHost()) {
+                uri = NamespaceFilter.insertNamespace(uri, namespace);
+                if (s3Config.isSignNamespace())
+                    resource = "/" + namespace + resource; // prepend to resource path for signing
+            } else {
+                // issue warning if namespace is specified and vhost is disabled because we can't put the namespace in the URL
+                log.warn("vHost namespace is disabled, so there is no way to specify a namespace in a pre-signed URL");
+            }
+        }
+        // insert bucket in host or path
+        uri = BucketFilter.insertBucket(uri, request.getBucketName(), s3Config.isUseVHost());
+        Map<String, String> parameters = new TreeMap();
+        if (request.getVersionId() != null) parameters.put(S3Constants.PARAM_VERSION_ID, request.getVersionId());
+        Map<ResponseHeaderOverride, String> headerOverrides = request.getHeaderOverrides();
+        for (ResponseHeaderOverride override : headerOverrides.keySet()) {
+            parameters.put(override.getQueryParam(), headerOverrides.get(override));
+        }
+
+        Map<String, List<Object>> headers = request.getHeaders();
+        String method = request.getMethod().name();
+
+        String date = null;
+        String serviceType = getServiceType();
+        if (headers.containsKey(S3Constants.AMZ_DATE)) {
+            date = RestUtil.getFirstAsString(headers, S3Constants.AMZ_DATE);
```

Comment:
`getDate()` already checks for `AMZ_DATE`, so why are we checking again here?

Reply:
Removed the first check.